energy-integration/mex-governance/src/vote.rs

```diff
@@ -66,16 +72,115 @@ pub trait VoteModule:
             .update(|sum| *sum += &user_energy);
         user_votes_in_week_mapper.set(&farm_votes);
 
+        self.update_top_farms(voting_week, &farm_votes_event);
+
         self.emit_vote_event(voting_week, farm_votes_event);
     }
 
+    fn update_top_farms(&self, week: usize, new_votes: &ManagedVec<FarmEmission<Self::Api>>) {
+        let mut top_farms = self.farm_emissions_for_week(week).get();
+        let mut redistributed_votes = self.redistributed_votes_for_week(week).get();
+
+        self.add_new_votes(week, &mut top_farms, new_votes);
+        self.selection_sort_farms_desc(&mut top_farms);
+        self.trim_top_farms(week, &mut top_farms, &mut redistributed_votes);
+
+        self.farm_emissions_for_week(week).set(&top_farms);
+        self.redistributed_votes_for_week(week)
+            .set(&redistributed_votes);
+    }
+
+    fn add_new_votes(
+        &self,
+        week: usize,
+        top_farms: &mut ManagedVec<FarmEmission<Self::Api>>,
+        new_votes: &ManagedVec<FarmEmission<Self::Api>>,
+    ) {
+        for farm_vote in new_votes.iter() {
+            let farm_address = farm_vote.farm_address.clone();
+            let vote_amount = farm_vote.farm_emission.clone();
+
+            let is_in_top_farms = self.top_farms_whitelist(week).contains(&farm_address);
+
+            if is_in_top_farms {
+                for i in 0..top_farms.len() {
+                    if &top_farms.get(i).farm_address == &farm_address {
+                        let mut farm_emission = top_farms.get(i);
+                        farm_emission.farm_emission += vote_amount;
+                        let _ = top_farms.set(i, &farm_emission);
+                        break;
+                    }
+                }
```

Comment:
This can be denested by extracting it in a function. Makes it a bit easier to read.

Reply:
Updated.